diff -Naur buildroot-2015.02/package/dvb-apps/dvb-apps.mk buildroot/package/dvb-apps/dvb-apps.mk
--- buildroot-2015.02/package/dvb-apps/dvb-apps.mk	2015-03-01 22:26:12.000000000 +0100
+++ buildroot/package/dvb-apps/dvb-apps.mk	2026-04-11 22:21:10.142592049 +0200
@@ -4,9 +4,9 @@
 #
 ################################################################################
 
-DVB_APPS_VERSION = 3d43b280298c39a67d1d889e01e173f52c12da35
-DVB_APPS_SITE = http://linuxtv.org/hg/dvb-apps
-DVB_APPS_SITE_METHOD = hg
+DVB_APPS_VERSION = 1.1.1+rev1500.orig
+DVB_APPS_SOURCE = linuxtv-dvb-apps_$(DVB_APPS_VERSION).tar.bz2
+DVB_APPS_SITE = http://ftp.debian.org/debian/pool/main/l/linuxtv-dvb-apps
 DVB_APPS_LICENSE = GPLv2 GPLv2+ LGPLv2.1+
 DVB_APPS_LICENSE_FILES = COPYING COPYING.LGPL
 
diff -Naur buildroot-2015.02/package/i2c-tools/i2c-tools.mk buildroot/package/i2c-tools/i2c-tools.mk
--- buildroot-2015.02/package/i2c-tools/i2c-tools.mk	2015-03-01 22:26:12.000000000 +0100
+++ buildroot/package/i2c-tools/i2c-tools.mk	2026-04-11 22:13:23.777573430 +0200
@@ -5,8 +5,8 @@
 ################################################################################
 
 I2C_TOOLS_VERSION = 3.1.1
-I2C_TOOLS_SOURCE = i2c-tools-$(I2C_TOOLS_VERSION).tar.bz2
-I2C_TOOLS_SITE = http://dl.lm-sensors.org/i2c-tools/releases
+I2C_TOOLS_SOURCE = i2c-tools-$(I2C_TOOLS_VERSION).tar.gz
+I2C_TOOLS_SITE = http://www.kernel.org/pub/software/utils/i2c-tools
 I2C_TOOLS_LICENSE = GPLv2+, GPLv2 (py-smbus)
 I2C_TOOLS_LICENSE_FILES = COPYING
 
diff -Naur buildroot-2015.02/package/linux-headers/Config.in.host buildroot/package/linux-headers/Config.in.host
--- buildroot-2015.02/package/linux-headers/Config.in.host	2015-03-01 22:26:12.000000000 +0100
+++ buildroot/package/linux-headers/Config.in.host	2026-04-11 22:13:23.777573430 +0200
@@ -67,6 +67,10 @@
 		bool "Linux 3.18.x kernel headers"
 		select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_18
 
+	config BR2_KERNEL_HEADERS_3_16
+		bool "Linux 3.16.x kernel headers"
+		select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_16
+
 	config BR2_KERNEL_HEADERS_VERSION
 		bool "Manually specified Linux version"
 endchoice
@@ -94,6 +98,18 @@
 
 config BR2_PACKAGE_HOST_LINUX_HEADERS_CUSTOM_3_17
 	bool "3.17.x"
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_17
+
+config BR2_PACKAGE_HOST_LINUX_HEADERS_CUSTOM_3_16
+	bool "3.16.x"
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_16
+
+config BR2_PACKAGE_HOST_LINUX_HEADERS_CUSTOM_3_15
+	bool "3.15.x"
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_15
+
+config BR2_PACKAGE_HOST_LINUX_HEADERS_CUSTOM_3_17
+	bool "3.17.x"
 	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_17
 
 config BR2_PACKAGE_HOST_LINUX_HEADERS_CUSTOM_3_16
diff -Naur buildroot-2015.02/package/pkgconf/pkgconf.mk buildroot/package/pkgconf/pkgconf.mk
--- buildroot-2015.02/package/pkgconf/pkgconf.mk	2015-03-01 22:26:12.000000000 +0100
+++ buildroot/package/pkgconf/pkgconf.mk	2026-04-11 22:13:23.777573430 +0200
@@ -5,7 +5,7 @@
 ################################################################################
 
 PKGCONF_VERSION = 0.8.9
-PKGCONF_SITE = http://rabbit.dereferenced.org/~nenolod/distfiles
+PKGCONF_SITE = http://sources.buildroot.net
 PKGCONF_SOURCE = pkgconf-$(PKGCONF_VERSION).tar.bz2
 PKGCONF_LICENSE = pkgconf license
 PKGCONF_LICENSE_FILES = COPYING
diff -Naur buildroot-2015.02/toolchain/toolchain-common.in buildroot/toolchain/toolchain-common.in
--- buildroot-2015.02/toolchain/toolchain-common.in	2015-03-01 22:26:12.000000000 +0100
+++ buildroot/toolchain/toolchain-common.in	2026-04-11 22:13:23.777573430 +0200
@@ -221,6 +221,18 @@
 	bool
 	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_17
 
+config BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_15
+	bool
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_14
+
+config BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_16
+	bool
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_15
+
+config BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_17
+	bool
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_16
+
 # This order guarantees that the highest version is set, as kconfig
 # stops affecting a value on the first matching default.
 config BR2_TOOLCHAIN_HEADERS_AT_LEAST
diff -Naur buildroot-2015.02/toolchain/toolchain-external/Config.in buildroot/toolchain/toolchain-external/Config.in
--- buildroot-2015.02/toolchain/toolchain-external/Config.in	2015-03-01 22:26:12.000000000 +0100
+++ buildroot/toolchain/toolchain-external/Config.in	2026-04-11 22:13:23.777573430 +0200
@@ -1054,6 +1054,18 @@
 	bool "3.15.x"
 	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_15
 
+config BR2_TOOLCHAIN_EXTERNAL_HEADERS_3_17
+	bool "3.17.x"
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_17
+
+config BR2_TOOLCHAIN_EXTERNAL_HEADERS_3_16
+	bool "3.16.x"
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_16
+
+config BR2_TOOLCHAIN_EXTERNAL_HEADERS_3_15
+	bool "3.15.x"
+	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_15
+
 config BR2_TOOLCHAIN_EXTERNAL_HEADERS_3_14
 	bool "3.14.x"
 	select BR2_TOOLCHAIN_HEADERS_AT_LEAST_3_14
